DiscordBot/Commands/ICommandProvider.cs

```diff
@@ -0,0 +1,11 @@
+﻿using System.Collections.Generic;
+
+namespace DiscordBot.Commands
+{
+    interface ICommandProvider
+    {
+        public Dictionary<string[], ICommand> Commands { get; set; }

```

Comment:
Isn't the command provider having a dictionary an implementation detail?
What you need from the interface is the ability to iterate all the commands, not necessarily get the whole dictionary.

Reply:
so remove it from the interface, and in the CommandProvider.cs make it private?

and also do a function GetCommands(), which just returns all the values of the dictionary?